src/runtime_src/tools/xclbinutil/RapidJsonUtilities.cxx

```diff
@@ -330,4 +362,161 @@ XclBinUtilities::read_cbor(std::istream& istr,
   }
 }
 
+void 
+XclBinUtilities::validate_against_schema(const std::string &nodeName, const rapidjson::Document & doc, const std::string & schema)
+{
+  rapidjson::Document sd;
+  if(sd.Parse(schema.c_str()).HasParseError()) {
+    XUtil::TRACE(boost::str(boost::format("Schema:\n %s") % schema));
+    throw std::runtime_error("Error: The given JSON schema is not valid JSON.");
+  }
+
+  rapidjson::SchemaDocument schemaDoc(sd);        // Convert to a schema document
+  rapidjson::SchemaValidator validator(schemaDoc);
+
+  // Valid the given JSON file
+  if (!doc.Accept(validator)) {
+    // The JSON image is invalid according to the schema
+    rapidjson::StringBuffer sb;
+    validator.GetInvalidSchemaPointer().StringifyUriFragment(sb);
+    
+    std::ostringstream buf;
+    buf << "Error: JSON schema violation" << std::endl;
+    boost::format simpleFormat("  %-22s: %s\n");
+    buf << simpleFormat % "JSON Node" % nodeName;
+    buf << simpleFormat % "Schema violation rule" % sb.GetString();
+    buf << simpleFormat % "Violation type" % validator.GetInvalidSchemaKeyword();
+
+    sb.Clear();
+    validator.GetInvalidDocumentPointer().StringifyUriFragment(sb);
+    buf << simpleFormat % "JSON document path" % sb.GetString();
+
+    XUtil::TRACE(boost::str(boost::format("Schema:\n %s") % schema));
+    XUtil::TRACE_PrintTree("JSON", doc);
+    throw std::runtime_error(buf.str());
+  }
+
+  XUtil::TRACE("JSON syntax successfully validated against the schema.");
+}
+
+
+static void recursive_collect_properties( const std::string& scope, rapidjson::Value::MemberIterator itrObject, XclBinUtilities::KeyTypeCollection& keyTypeCollection);
+static void recursive_collect_array( const std::string& scope, rapidjson::Value::MemberIterator itrObject, XclBinUtilities::KeyTypeCollection& keyTypeCollection);
+
+
+static 
+void recursive_collect_array( const std::string& scope, 
+                              rapidjson::Value::MemberIterator itrObject, 
+                              XclBinUtilities::KeyTypeCollection& keyTypeCollection)
+{
+  const rapidjson::Value::MemberIterator itemsItr = itrObject->value.FindMember("items");
+  if (itemsItr == itrObject->value.MemberEnd())
+    return;
+
+  const rapidjson::Value::MemberIterator typeItr = itemsItr->value.FindMember("type");
+  if (typeItr == itemsItr->value.MemberEnd()) 
+    return;
+
+  // -- array of objects --
+  if (std::string("object") == typeItr->value.GetString()) {
```

Comment:
early-exit ?
if (x != y)
  return;

But I see it doesn't fit the pattern of extension with value types so this probably shouldn't be changed.

Reply:
Currently this code only support an array of objects.   While not required for 2021.1, I am expected the need to support array of strings and integers.  When that should occur, we will have a set of conditions where early exits would not be possible.  

In other portions of the code, it is indeed written with early exits, but in this piece of code it isn't recommended.